src/awkward/operations/convert.py

```diff
@@ -2846,15 +2850,15 @@ def handle_arrow(obj):
                 return ak.layout.RecordArray(child_array, obj.schema.names)
 
         elif isinstance(obj, pyarrow.lib.Table):
-            chunks = []
-            for batch in obj.to_batches():
-                chunk = handle_arrow(batch)
-                if len(chunk) > 0:
-                    chunks.append(chunk)
-            if len(chunks) == 1:
-                return chunks[0]
+            batches = []
+            for record_batch in obj.combine_chunks().to_batches():
```

Comment:
`.combine_chunks()` combines (possible) chunks of a table by creating a new one. Not sure if this might be resulting in memory issues.
Overall worth to think about treating Tables like RecordBatches and iterate over columns rather than converting to batch obj. After all, it is a columnar memory format. As such, each column could be type checked separately early on.

Reply:
I've checked out your code and I'm playing with it now. As I understand from [the documentation](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.combine_chunks) and a few tests, `combine_chunks` ensures that you'll have zero or one RecordBatches, so Arrow does the concatenation and Awkward doesn't have to (line 2861).

If a table consists of exactly one RecordBatch (arrays are not ChunkedArrays):

```python
>>> single = pyarrow.Table.from_batches([
...     pyarrow.RecordBatch.from_arrays([ak.to_arrow(ak.Array(range(10)))], ["name"])
... ])
```

then calling `combine_chunks` on it does not allocate a new array. Arrow is smart enough to notice that there's nothing to concatenate, so it returns a view to the original.

```python
>>> np.frombuffer(single.to_batches()[0].column(0).buffers()[1], np.int64).ctypes.data
94720844398896
>>> np.frombuffer(single.combine_chunks().to_batches()[0].column(0).buffers()[1], np.int64).ctypes.data
94720844398896
```

This is not the case when there's multiple RecordBatches:

```python
>>> multi = pyarrow.Table.from_batches([
...     pyarrow.RecordBatch.from_arrays([ak.to_arrow(ak.Array(range(10)))], ["name"]),
...     pyarrow.RecordBatch.from_arrays([ak.to_arrow(ak.Array(range(10)))], ["name"]),
...     pyarrow.RecordBatch.from_arrays([ak.to_arrow(ak.Array(range(10)))], ["name"]),
... ])
>>> np.frombuffer(multi.to_batches()[0].column(0).buffers()[1], np.int64).ctypes.data
94720844573024
>>> np.frombuffer(multi.combine_chunks().to_batches()[0].column(0).buffers()[1], np.int64).ctypes.data
140673485410304
```

So I'm going to change the code to combine the chunks, check to see if the result has zero or one RecordBatches, and if it's one, just transform that without a concatenation.